### PR TITLE
[GH-57] add autocomplete

### LINKED ISSUE #57: Add slash command autocomplete functionality
From 5.24 MM server has a slash command autocomplete functionality feature. An API can be found [here](https://github.com/mattermost/mattermost-server/blob/master/model/command_autocomplete.go#L28)
To implement the feature add `AutocompleteData` struct to the `Command` when registering a slash command.

Similar PRs:
For Todo plugin https://github.com/mattermost/mattermost-plugin-todo/pull/69
For Agenda plugin https://github.com/mattermost/mattermost-plugin-agenda/pull/40

### PR DESCRIPTION
**Pull request by jfrerich**
#### Summary
This PR adds autocomplete to the confluence plugin.  

A new endpoint is added to handle the dynamic retrieving of subscription aliases for the `edit` and `unsubscribe` commands.

#### Ticket Link
Fixes #57

### CODE REVIEW

File: server/controller/get_subscriptions.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+var autocompleteGetChannelSubscriptions = &Endpoint{
+	RequiresAdmin: true,
+	Path:          "/autocomplete/GetChannelSubscriptions",
+	Method:        http.MethodGet,
+	Execute:       handleGetChannelSubscriptions,
+}
+
+func handleGetChannelSubscriptions(w http.ResponseWriter, r *http.Request) {
+	channelID := r.FormValue("channel_id")
```
Comment (jfrerich) [inline server/controller/get_subscriptions.go:19]: @iomodo, I'm not sure of the exact commit needed from master, but I believe this is the functionality that is not currently in a released version?

File: server/command/command.go [hunk 2]
```diff
@@ -79,9 +79,43 @@ func GetCommand() *model.Command {
 		AutoComplete:     true,
 		AutoCompleteDesc: "Available commands: subscribe, list, unsubscribe \"<name>\", edit \"<name>\", install cloud/server, help.",
 		AutoCompleteHint: "[command]",
+		AutocompleteData: getAutoCompleteData(),
 	}
 }
 
+func getAutoCompleteData() *model.AutocompleteData {
+	confluence := model.NewAutocompleteData("confluence", "[command]", "Available commands:  subscribe, list, unsubscribe \"<name>\", edit \"<name>\", install cloud/server, help")
```
Comment (mickmister) [inline server/command/command.go:87]: Nit: Is there supposed to be two spaces here?

```suggestion
	confluence := model.NewAutocompleteData("confluence", "[command]", "Available commands: subscribe, list, unsubscribe \"<name>\", edit \"<name>\", install cloud/server, help")
```

File: server/controller/get_subscriptions.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	out := []model.AutocompleteListItem{}
+	for alias := range subscriptions {
+		out = append(out, model.AutocompleteListItem{
+			Item: alias,
+		})
+	}
+
+	b, _ := json.Marshal(out)
+	w.Header().Set("Content-Type", "application/json")
+	_, _ = w.Write([]byte(string(b)))
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply server/controller/get_subscriptions.go:35]: Will change this.  I was grabbing from the existing code in get_subscription.go

File: server/controller/get_subscriptions.go (hunk 1 above)
Comment (jfrerich) [reply server/controller/get_subscriptions.go:19]: There was a change to autocomplete that now passes the `channel_id` to the URL as a query param.  This was only recently introduced.  Without this new feature, you have no way of detecting the current channel for which you are typing an autocomplete command.

File: server/controller/get_subscriptions.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	out := []model.AutocompleteListItem{}
+	for alias := range subscriptions {
+		out = append(out, model.AutocompleteListItem{
+			Item: alias,
+		})
+	}
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply server/controller/get_subscriptions.go]: I could not access `sub.Alias` directly, but added new interface method to get a subscription Alias.  See latest commit!

File: server/controller/get_subscriptions.go (hunk 1 above)
Comment (iomodo) [reply server/controller/get_subscriptions.go:19]: True, changes are for 5.26 so you can grab the lates master for now.

Comment (mickmister) [reply server/controller/get_subscriptions.go:19]: @jfrerich @iomodo Do we need to update the minimum server version because of this change?

Comment (jfrerich) [reply server/controller/get_subscriptions.go:19]: Yes, we needed to update the min server version. I committed the `min_server_version` to `5.26.0`.  

What is the correct action with respect to merging? I added the `Do Not Merge/Awaiting PR` label.  We can merge post `5.26.0` release.  How does that sound and is that necessary?

Comment (jfrerich) [reply server/controller/get_subscriptions.go:19]: @iomodo, @mickmister 

* updated `min_server_version` to `v5.26.0` in plugin.json
* updated go.mod with latest server version commit via `go get github.com/mattermost/mattermost-server/v5@93a537a636299e1e049505bdaf072f8d3ae974a4`

File: go.mod [hunk 5]
```diff
@@ -5,7 +5,7 @@ go 1.13
 require (
 	bou.ke/monkey v1.0.2
 	github.com/gorilla/mux v1.7.4
-	github.com/mattermost/mattermost-server/v5 v5.24.1
+	github.com/mattermost/mattermost-server/v5 v5.3.2-0.20200717105608-93a537a63629
```
Comment (hanzei) [inline go.mod:8]: What changes in `mattermost-server` that are only in `master` are needed for this PR?

Comment (iomodo) [reply go.mod:8]: getting the `channel_id` in the dynamic argument endpoint would be the one.

Comment (hanzei) [reply go.mod:8]: Got it :+1:

**Pull request merged**